src/shell.ts

```diff
@@ -0,0 +1,175 @@
+import { ChildProcess, spawn } from "child_process"
+import { randomUUID } from "crypto"
+import { readFile } from "fs/promises"
+import { Readable as ReadableStream } from "stream"
+
+import { redact } from "./format"
+import { Logger } from "./logger"
+
+export const displayShellCommand = (
+  execPath: string,
+  args: string[],
+  {
+    itemsToRedact,
+  }: {
+    itemsToRedact?: string[]
+  } = {},
+) => {
+  return redact(`${execPath} ${args.join(" ")}`, itemsToRedact ?? [])
+}
+
+export const tryReadFile = async (...args: Parameters<typeof readFile>) => {
+  try {
+    return (await readFile(...args)).toString().trim()
+  } catch (error) {
+    if (
+      /*
+      Test for the following error:
+        [Error: ENOENT: no such file or directory, open '/foo'] {
+          errno: -2,
+          code: 'ENOENT',
+          syscall: 'unlink',
+          path: '/foo'
+        }
+      */
+      !(error instanceof Error) ||
+      (error as { code?: string })?.code !== "ENOENT"
+    ) {
+      throw error
+    }
+  }
+}
+
+type ShellCommandRunnerConfiguration = {
+  itemsToRedact?: string[]
+  shouldTrackProgress?: boolean
+  cwd?: string
+  onChild?: (child: ChildProcess) => void
+}
+export const getShellCommandRunner = (
```

Comment:
just to double-check - do we include only methods that are reused in more than 1 project? or it could be that the method is used only in gitspiegel ?

Reply:
> just to double-check - do we include only methods that are reused in more than 1 project

"Used in more than 1 project" is a good indicator that it should be moved to this repository _but_ not a rule. For some new feature it might make sense to submit it here directly.

> or it could be that the method is used only in gitspiegel ?

Most projects have some sort of command runner:

https://github.com/paritytech/command-bot/blob/c4da170e1eac6f49a532076efb96348e2298024e/src/shell.ts#L26

https://github.com/paritytech/bench-bot/blob/cd3b2943d911ae29e41fe6204788ef99c19412c3/bench.js#L35

https://github.com/paritytech/github-issue-sync/blob/054ad6cbb1f3ff1598577bca9f94aa32e4b9116d/src/command.ts#L36

This `getShellCommandRunner` would replace them